Feature/12 see nft details

Goddak opened the pull request:
Build out the NFT Detail view.

Fixes #12 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Noteworthy changes
 - Added `REACT_APP_IPFS_GATEWAY_URL=https://ipfs.io/ipfs/` to `.env`.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

filipmanole commented on src/components/NftCard/NftCard.scss:
```diff
@@ -52,4 +52,4 @@
         }
     }
-}
+}
```
```suggestion
}

```

filipmanole replied:
no new line at EOF

filipmanole commented on src/widgets/NftDetailWidget/subcomponents/NftDetailActivity/NftDetailActivity.scss:
```diff
@@ -0,0 +1,16 @@
+			margin-left: 1rem;
+		}
+	}
+}
```
```suggestion
}

```

filipmanole replied:
no new line at EOF

filipmanole commented on src/widgets/NftDetailWidget/subcomponents/NftDetailPortrait/NftDetailPortrait.tsx:
```diff
@@ -4,7 +4,7 @@ import './NftDetailPortrait.scss';
 interface CollectionPortraitProps {
   backgroundImage: string;
-  className?: string;
+  className: string;
```
Why is it mandatory to have the className parameter?

Goddak replied:
Because it has a default value so is never undefined.
We can make it optional, I don't mind :man_shrugging:

filipmanole commented on src/widgets/NftDetailWidget/NftDetailWidget.tsx:
```diff
@@ -1,31 +1,128 @@
+                defaultOpen
+              />
+            </div>
+            {/* <div className="nft-detail-widget__activities">
```
commented code with no description

filipmanole commented on src/widgets/NftDetailWidget/NftDetailWidget.tsx:
```diff
@@ -1,31 +1,128 @@
+                defaultOpen
+              />
+            </div>
+            {/* <div className="nft-detail-widget__activities">
+              <NftDetailAccordian
+                label="Item Activity"
+                content={(
+                  <NftDetailActivity logs={eventLogs} />
+                )}
+              />
+            </div>
+            <div className="nft-detail-widget__details">
+              <NftDetailAccordian
+                label="Details"
+                content={(
+                  <p>{nftMetadata?.description}</p>
+                )}
+              />
+            </div> */}
```
commented code with no description

filipmanole commented on src/widgets/NftDetailWidget/NftDetailWidget.tsx:
```diff
@@ -1,31 +1,128 @@
+
+  const { collectionImage, collectionToken } = config;
+  const { tokensData } = collection;
+  // const { eventLogs } = token;
```
commented code

filipmanole commented on src/redux/stores/token/tokenApi.ts:
```diff
@@ -0,0 +1,56 @@
+    }
+
+
+    // try {
+    //   const contract = getCollectionContract();
+    //   if (!contract) return;
+    //   contract.on('Transfer', (from, to, tokenId, event) => {
+    //     console.log('Transfer Event:', from, to, tokenId, event);
+    //   });
+    // } catch (err) {
+    //   // TODO
+    //   console.log('contract.on Transfer CATCH', err);
+    // }
```
commented code with no description

Goddak replied:
![image](https://user-images.githubusercontent.com/12832288/216109406-4a89dcda-fd19-41ce-a28b-2b2d9e16020e.png)

filipmanole commented on src/widgets/NftDetailWidget/NftDetailWidget.tsx:
```diff
@@ -1,31 +1,128 @@
+import useNftMetadata from '../../hooks/useNftMetadata';
+import { useAppDispatch, useAppSelector } from '../../redux/hooks';
+import { fetchNFTMetadata, fetchNFTMetadataParams } from '../../redux/stores/collection/collectionApi';
+// import { fetchNFTActivity } from '../../redux/stores/token/tokenApi';
+import { AppRoutes } from '../../routes';
+import NftDetailAccordian from './subcomponents/NftDetailAccordian/NftDetailAccordian';
+// import NftDetailActivity from './subcomponents/NftDetailActivity/NftDetailActivity';
+// import NftDetailAttributeCard from './subcomponents/NftDetailAttributeCard/NftDetailAttributeCard';
```
commented code

piersss commented on src/widgets/NftDetailWidget/NftDetailWidget.tsx:
```diff
@@ -1,31 +1,100 @@
+              />
+            </div>
+          </div>
+          <Link to={`/${AppRoutes.swap}/${selectedTokenId}`}>
+            <Button text="Proceed to buy" className="nft-detail-widget__proceed-button" />
+          </Link>
```
Don't put a `button` in an `a`. Use `@extend %button-is-primary`.

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailAccordian/NftDetailAccordian.scss:
```diff
@@ -0,0 +1,45 @@
+        padding-right: .5rem;
+    }
+    
+    &__state-indicator {
+        transition: transform .3s ease-out;
+        transform-origin: center;
+    }
+
+    &__content {
+        height: 0;
+        transition: height .3s;
+        overflow: hidden;
+        will-change: transform;
+
+        &--is-open {
+            height: 100%;
+        }
+    }
+    @include for-size(tablet-landscape-up) {
+        &__state-indicator {
+            display: none;
+        }
+
+        &__content {
+            height: 100%;
+        }
+    }
+}
```
```suggestion
    &__state-indicator {
        transition: transform .3s ease-out;
        transform-origin: center;
        
         @include for-size(tablet-landscape-up) {
              display: none;
         }
    }

    &__content {
        height: 0;
        transition: height .3s;
        overflow: hidden;
        will-change: transform;
        
        @include for-size(tablet-landscape-up) {
            height: 100%;
         }

        &--is-open {
            height: 100%;
        }
    }
}
```

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailActivity/NftDetailActivity.scss:
```diff
@@ -0,0 +1,16 @@
+@import "src/styles/index";
+
+.nft-detail-widget {
```
```suggestion
.nft-detail-activity {
```

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailActivity/NftDetailActivity.scss:
```diff
@@ -0,0 +1,16 @@
+		&:last-child {
+			margin-bottom: 0;
+		}
+		p {
```
Don't style html tags directly

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailAttributeCard/NftDetailAttributeCard.scss:
```diff
@@ -0,0 +1,29 @@
+        font-weight: var(--fw-700);
+    }
+
+    @include for-size(tablet-portrait-up) {
+        font-size: 1rem;
+    }
```
```suggestion
```

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailAttributeCard/NftDetailAttributeCard.scss:
```diff
@@ -0,0 +1,29 @@
+    padding: 1em;
+    background: #1A1E25;
+    border-radius: 4px;
+
```
```suggestion
    @include for-size(tablet-portrait-up) {
        font-size: 1rem;
    }
```

piersss commented on package.json:
```diff
@@ -21,6 +22,7 @@
     "@web3-react/core": "^6.1.9",
     "@web3-react/injected-connector": "^6.0.7",
     "@web3-react/walletconnect-connector": "6.2.13",
+    "axios": "^1.2.3",
```
Axios is not used

piersss reviewed (changes requested):
Looking good so far.

1. I think you should make a new store for nft-details. I understand using the collection store has it's benefits but you should also think about error handling (like 404) and the activity log data. Where are you going to leave that? Also in the collection store? It's best to keep it separate.

2. All loading states should be done in store. When you do everything in the nft-details store you won't need to do any loading states in component anyway. `useNftMetadata` won't be needed either then.

3. You have no logic for an unconnected wallet. Right now it just says "loading". You could maybe use the solution I made in `CollectionWidget`. Although there's no design for it yet so just use a "404" placeholder for it now and we'll make a separate ticket.

4. Using browser back button won't load new data.

5. Airswap protocols just released new version with new getToken functions. Please. use those. https://github.com/airswap/airswap-protocols/pull/1000

Goddak commented:
> Looking good so far.
> 
> 1. I think you should make a new store for nft-details. I understand using the collection store has it's benefits but you should also think about error handling (like 404) and the activity log data. Where are you going to leave that? Also in the collection store? It's best to keep it separate.

I had made a `token` store? it had an `eventLogs` property for storing the activity log data? Is that not what you meant?
I've renamed it now to `nftDetail`.

> 2. All loading states should be done in store. When you do everything in the nft-details store you won't need to do any loading states in component anyway. `useNftMetadata` won't be needed either then.

OK, I can move those into the store ..... :magic_wand: BAM it's done, magic :rofl:

> 3. You have no logic for an unconnected wallet. Right now it just says "loading". You could maybe use the solution I made in `CollectionWidget`. Although there's no design for it yet so just use a "404" placeholder for it now and we'll make a separate ticket.

:magic_wand: BAM it's done, magic :rofl:

> 4. Using browser back button won't load new data.

This seems to be working fine for me? Maybe I've fixed it whilst moving stuff to store today?

> 5. Airswap protocols just released new version with new getToken functions. Please. use those. [airswap/airswap-protocols#1000](https://github.com/airswap/airswap-protocols/pull/1000) . But @filipmanole is also working on that so sync up with him.

Had a brief chat via Discord with @filipmanole he wasn't quite there and neither was I, we will liaise to make sure this is only done once.

piersss commented:
```
I had made a token store? it had an eventLogs property for storing the activity log data? Is that not what you meant?
I've renamed it now to nftDetail.
```

Oh yeah, totally missed that sorry. Right on 👍 

```
This seems to be working fine for me? Maybe I've fixed it whilst moving stuff to store today?
```

It didnt work when you go from `nft-detail/3` to `nft-detail/2`. This could happen when user enters id's manually in address bar and then go back and forth between them.

piersss commented on .env.example:
```diff
@@ -11,4 +11,7 @@ REACT_APP_COLLECTION_IMAGE=collection/collection-example-image.png
+REACT_APP_RPC_URL_5=
+
+REACT_APP_IPFS_GATEWAY_URL=https://ipfs.io/ipfs/
+REACT_APP_PAYMENT_TOKEN=AST
```
We already have REACT_APP_PAYMENT_TOKEN for this.

piersss commented on src/components/NftCard/NftCard.tsx:
```diff
@@ -22,7 +24,7 @@ const NftCard = ({
     <div className="nft-card__info-wrapper">
       <h3 className="nft-card__name">{name}</h3>
-      <h4 className="nft-card__price">{`${price} ETH`}</h4>
+      <h4 className="nft-card__price">{`${format(BigNumber.from(price))} ${process.env.REACT_APP_PAYMENT_TOKEN}`}</h4>
```
Get token symbol from currencyToken in metadata store

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailAccordian/NftDetailAccordian.tsx:
```diff
@@ -0,0 +1,43 @@
+import React, { FC, useRef, useState } from 'react';
```
Make Accordion component in /src/components so we can reuse this.

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailAccordian/NftDetailAccordian.tsx:
```diff
@@ -0,0 +1,43 @@
+import './NftDetailAccordian.scss';
+
+interface NftDetailAccordianProps {
+  label: string;
+  content: string | JSX.Element;
+  defaultOpen?: boolean;
+  className?: string;
```
```suggestion
  isDefaultOpen?: boolean;
  content: string | JSX.Element;
  label: string;
  className?: string;
```

piersss replied:
```suggestion
  isDefaultOpen?: boolean;
  content: string | JSX.Element;
  label: string;
  className?: string;
```

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailAccordian/NftDetailAccordian.tsx:
```diff
@@ -0,0 +1,43 @@
+}
+
+const NftDetailAccordian: FC<NftDetailAccordianProps> = ({
+  label, content, defaultOpen = false, className = '',
```
```suggestion
  isDefaultOpen: false,
  content,
  label,
  className = '',
```

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailActivity/NftDetailActivity.scss:
```diff
@@ -0,0 +1,17 @@
+
+.nft-detail-activity {
+	&__activity {
+		border: .1rem #696969 solid;
```
Use color css variables

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailActivity/NftDetailActivity.scss:
```diff
@@ -0,0 +1,17 @@
+
+.nft-detail-activity {
+	&__activity {
+		border: .1rem #696969 solid;
+		padding: 1rem;
+		display: flex;
+		margin-bottom: 1rem;
```
https://github.com/airswap/airswap-marketplace/blob/main/CONTRIBUTING.md#order-of-properties

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailAttributeCard/NftDetailAttributeCard.scss:
```diff
@@ -0,0 +1,28 @@
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+    padding: 1em;
```
```suggestion
    padding: 1rem;
```

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailAttributeCard/NftDetailAttributeCard.scss:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    &__label {
+        color: #9EA3AC;
```
Use color css variables

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailSaleInfo/NftDetailSaleInfo.scss:
```diff
@@ -0,0 +1,14 @@
+@import "src/styles/index";
+
+.nft-detail-sale-info {
+    margin-bottom: 1.2rem;
```
https://github.com/airswap/airswap-marketplace/blob/main/CONTRIBUTING.md#rem

piersss replied:
```suggestion
    margin-bottom: 1.25rem;
```

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailSaleInfo/NftDetailSaleInfo.scss:
```diff
@@ -0,0 +1,14 @@
+    margin-bottom: 1.2rem;
+
+    &__price {
+        font-size: 1.2rem;
```
https://github.com/airswap/airswap-marketplace/blob/main/CONTRIBUTING.md#rem

piersss replied:
```suggestion
        font-size: 1.25rem;
```

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailAttributes/NftDetailAttributes.tsx:
```diff
@@ -0,0 +1,19 @@
+import React, { FC } from 'react';
+
+import { Attribute } from '../../../../redux/stores/nftDetail/nftDetailSlice';
```
This has to come from `entites/CollectionToken`. But that part @filipmanole is still working on, so it's ok to leave it for now.

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailAccordian/NftDetailAccordian.tsx:
```diff
@@ -0,0 +1,43 @@
+    <div className={`nft-detail-accordian ${className}`}>
+      <div className="nft-detail-accordian__header" ref={headerRef}>
+        <p className="nft-detail-accordian__label">{label}</p>
+        <p className="nft-detail-accordian__state-indicator" style={{ transform: `rotate(${stateIndicator}deg)` }}>V</p>
```
`stateIndicator` not needed. Use `Icon` component for icons. Icon name is 'chevron-down'.

```suggestion
        <p className="nft-detail-accordian__state-indicator" style={{ transform: `rotate(${isOpen ? 0 : 180}deg)` }}>V</p>
```

piersss commented on src/widgets/NftDetailWidget/NftDetailWidget.scss:
```diff
@@ -35,6 +16,22 @@
     }
 
+    &__main-info {
+        &--tablet-only {
+            display: none;
+
+            p {
+                margin-bottom: 0;
+            }
+        }
```
If you want to hide things on mobile and show on tablet do this:

```suggestion
     &__main-info {
        display: none;

        @include for-size(tablet-portrait-up) {
            display: block;
        }
    }
```

Goddak replied:
Poorly named modifier, but it's a separate behaviour to the default. It's easier to see with the current changes. My brain is tired, I'll have a ponder re how to update/replace this.

piersss replied:
We have the `for-size` mixin for responsive styles. Let's not mix things up.

piersss commented on src/widgets/NftDetailWidget/NftDetailWidget.scss:
```diff
@@ -43,7 +40,65 @@
+    }
+
+    @include for-size(tablet-landscape-up) {
+        display: grid;
```
Just use `display: flex`. This is way too complicated for just 2 columns.

piersss replied:
Ok better leave it for now then!

piersss commented on src/widgets/NftDetailWidget/NftDetailWidget.scss:
```diff
@@ -43,7 +40,65 @@
+        }
+    }
+
+    @include for-size(tablet-landscape-up) {
```
This should be above all children. All children should have it's own media query. Look at other components like `WalletConnector.scss` how it's done.

piersss commented on src/widgets/NftDetailWidget/subcomponents/ConnectedNftDetailWidget/ConnectedNftDetailWidget.tsx:
```diff
@@ -0,0 +1,87 @@
+    getDataForSelectedId();
+  }, [selectedTokenId]);
+
+  return (
+    isLoading
+    ? <p>Loading</p>
+    : (
+      <div className="nft-detail-widget">
```
```suggestion
  if (isLoading) {
    return (
      <div className={`nft-detail-widget ${className}`}>
        Loading
      </div>
    );
  }

  return (
    <div className={`nft-detail-widget ${className}`}>
```

piersss replied:
```suggestion
  if (isLoading) {
    return (
      <div className={`nft-detail-widget ${className}`}>
        Loading
      </div>
    );
  }

  return (
    <div className={`nft-detail-widget ${className}`}>
```

piersss commented:
Tested with `0x33fd426905f149f8376e227d0c9d3340aad17af1` collection token.

![Screenshot 2023-02-13 at 08 01 18](https://user-images.githubusercontent.com/86911296/218392775-122c80e4-5d12-439e-baed-5ddc96331e6c.png)

- The portrait `div` should be an `img` so the image sizes correctly.
- On desktop the attributes should be on the left.

piersss commented on src/components/Accordion/Accordion.tsx:
```diff
@@ -0,0 +1,42 @@
+  useEventListener('click', onClick, headerRef);
+  return (
+    <div className={`accordion ${className}`}>
+      <div className="accordion__header" ref={headerRef}>
```
Header should be button

piersss commented on src/redux/stores/nftDetail/nftDetailSlice.ts:
```diff
@@ -0,0 +1,98 @@
+  message?: string;
+}
+
+export interface Attribute {
+  'trait_type': string;
+  value: string;
+}
+
+export interface TokenInfoMetadata {
+  attributes: Array<Attribute>
+  description: string
+  edition: number
+  image: string
+  name: string
+}
+
+export interface TokenMeta {
+  id: number;
+  name: string;
+  image: string;
+  description: string;
+  attributes: Array<Attribute>;
+  price: string;
+  symbol: string;
+}
```
```suggestion
```

This data can be obtained from `CollectionToken` in @filipmanole branch.

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailAttributeCard/NftDetailAttributeCard.scss:
```diff
@@ -0,0 +1,29 @@
+    &__label {
+        color: var(--c-text-grey);
+        font-size: .75rem;
+        font-weight: var(--fw-400);
```
```suggestion
        font-weight: var(--fw-400);
        
        @include for-size(tablet-portrait-up) {
            font-size: .875rem;
        }
```

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailAttributeCard/NftDetailAttributeCard.scss:
```diff
@@ -0,0 +1,29 @@
+
+    &__value {
+        font-size: .75rem;
+        font-weight: var(--fw-700);
```
```suggestion
        font-weight: var(--fw-700);
        
        @include for-size(tablet-portrait-up) {
            font-size: .875rem;
        }
```

piersss commented on src/widgets/NftDetailWidget/subcomponents/NftDetailMainInfo/NftDetailMainInfo.scss:
```diff
@@ -1,18 +1,13 @@
 .nft-detail-main-info {
+    &__owner {
+        color: var(--c-text-grey);
+        font-size: .8rem;
```
```suggestion
        font-size: .8125rem;
```

piersss reviewed (changes requested):
Content is hidden by buy button:

![Screenshot 2023-02-18 at 22 17 31](https://user-images.githubusercontent.com/86911296/219900415-3d548f90-a551-4b0f-9b63-ff0a18e707a0.png)

Desktop needs improvement:

- Traits should be on the bottom left
- Border around details section
- Proceed to buy button needs to be narrower
- Margin under portrait should be bigger

![Screenshot 2023-02-18 at 23 01 06](https://user-images.githubusercontent.com/86911296/219901413-b1cfbc73-e379-44e5-a620-efd4b182bc70.png)

Navigating doesn't work properly:

https://user-images.githubusercontent.com/86911296/219901647-6bc09bb3-4ac1-4c75-b28e-ca619092ddc9.mov

piersss commented on src/redux/stores/nftDetail/nftDetailSlice.ts:
```diff
@@ -0,0 +1,86 @@
+  type: string;
+}
+
+export interface Error {
```
This will be confused with the javascript Error object. In airswap-web we use AppError with enums for each error so it's more easy to translate.

piersss replied:
Name it `NftDetailError` for now and we will do an Error refactor later.

Goddak commented on src/components/Accordion/Accordion.tsx:
```diff
@@ -0,0 +1,42 @@
+  useEventListener('click', onClick, headerRef);
+  return (
+    <div className={`accordion ${className}`}>
+      <button className="accordion__header focus:outline-none" ref={headerRef} type="button">
```
I have solved this in my last commit, didn't see this message before hand though so it's still a button. Do you feel strongly about the H2?

piersss replied:
Yeah it's semantically incorrect.

piersss reviewed (dismissed):
You forgot some Accordion feedback in my previous review.

Goddak commented on src/components/Accordion/Accordion.scss:
```diff
@@ -0,0 +1,45 @@
+        font-size: .9rem;
+        font-weight: var(--fw-700);
+        padding-bottom: .75rem;
+    }
```
`@extend %button-border-style;` doesn't work here, these aren't traditional buttons

piersss reviewed (approved):
LGTM
